File: java/src/test/java/com/example/hexagonal/usecase/ProductServiceTest.java
package com.example.hexagonal.usecase;

import static com.example.hexagonal.domain.ModelFixture.DEFAULT_ID;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;

import com.example.hexagonal.domain.ModelFixture;
import com.example.hexagonal.domain.Product;
import com.example.hexagonal.port.out.ProductJpaRepository;
import com.example.hexagonal.usecase.dto.DtoFixture;
import com.example.hexagonal.usecase.dto.ProductDto;
import java.util.Optional;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

@SpringBootTest
class ProductServiceTest {

  @Autowired ProductService sut;
  @MockBean ProductJpaRepository mockRepository;

  @Test
  void getProduct() {
    final Product expected = ModelFixture.aProduct();

    Mockito
        .when(mockRepository.findById(anyLong()))
        .thenReturn(Optional.of(expected));

    final Product actual = sut.getProduct(DEFAULT_ID);

    assertThat(actual).isEqualTo(expected);
  }

  @Test
  void updateProduct() {
    final Product expected = ModelFixture.aProduct();

    Mockito
        .when(mockRepository.findById(anyLong()))
        .thenReturn(Optional.of(expected));
    Mockito
        .when(mockRepository.save(any(Product.class)))
        .thenReturn(expected);

    final ProductDto dto = DtoFixture.aProductDtoOf(DEFAULT_ID);
    dto.setName("changed");
    dto.setStock(2);

    assertDoesNotThrow(() -> {
      sut.updateProduct(DEFAULT_ID, dto);
    });
  }
}

File: kotlin/src/test/kotlin/com/example/hexagonal/usecase/ProductServiceTest.kt
package com.example.hexagonal.usecase

import com.example.hexagonal.TestConfig
import com.example.hexagonal.domain.Product
import com.example.hexagonal.port.driven.MessagePort
import com.example.hexagonal.port.driven.ProductJpaRepository
import com.example.hexagonal.usecase.dto.ProductDto
import io.kotest.assertions.throwables.shouldNotThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.extensions.spring.SpringExtension
import io.kotest.matchers.shouldBe
import io.mockk.every
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.TestConstructor
import java.util.Optional

@SpringBootTest
@Import(TestConfig::class)
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
class ProductServiceTest(
    sut: ProductService,
    mockRepository: ProductJpaRepository,
    mockMessagePort: MessagePort,
) : DescribeSpec() {
    override fun extensions() = listOf(SpringExtension)

    init {
        describe("When calls 'getProduct' api") {
            it("should returns an entity") {
                every { mockRepository.findById(any<Long>()) } returns Optional.of(fixture)

                val actual = sut.getProduct(fixture.id)

                actual.id shouldBe fixture.id
            }
        }

        describe("When calls 'updateProduct' api") {
            it("should now throw exception") {
                every { mockRepository.findById(any<Long>()) } returns Optional.of(fixture)
                every { mockRepository.save(any()) } returns fixture
                every { mockMessagePort.send(any()) } returns Unit

                val dto = ProductDto.aProductDtoOf(fixture.id)
                    .apply {
                        this.name = "changed"
                        this.stock = 2
                    }

                shouldNotThrow<Exception> {
                    sut.updateProduct(fixture.id, dto)
                }
            }
        }
    }

    val fixture: Product = Product.aProduct()
}
